sklearn/cluster/_hdbscan/tests/test_hdbscan.py

```diff
@@ -0,0 +1,329 @@
+    n_clusters = len(set(labels)) - int(-1 in labels)
+    assert n_clusters == n_clusters_true
+
+    ALGOS_TREES = {
+        "kdtree": KDTree,
+        "balltree": BallTree,
+    }
+    metric_params = {
+        "mahalanobis": {"V": np.eye(X.shape[1])},
+        "seuclidean": {"V": np.ones(X.shape[1])},
+        "minkowski": {"p": 2},
+        "wminkowski": {"p": 2, "w": np.ones(X.shape[1])},
+    }.get(metric, None)
+    if algo not in ("auto", "brute"):
+        if metric not in ALGOS_TREES[algo].valid_metrics:
+            with pytest.raises(ValueError):
+                HDBSCAN(
+                    algorithm=algo,
+                    metric=metric,
+                    metric_params=metric_params,
+                ).fit(X)
+        elif metric == "wminkowski":
+            with pytest.warns(FutureWarning):
+                HDBSCAN(
+                    algorithm=algo,
+                    metric=metric,
+                    metric_params=metric_params,
+                ).fit(X)
+        else:
+            HDBSCAN(
+                algorithm=algo,
+                metric=metric,
+                metric_params=metric_params,
+            ).fit(X)
+
+
+def test_hdbscan_dbscan_clustering():
+    clusterer = HDBSCAN().fit(X)
+    labels = clusterer.dbscan_clustering(0.3)
+    n_clusters = len(set(labels)) - int(-1 in labels)
+    assert n_clusters == n_clusters_true
+
+
+def test_hdbscan_high_dimensional():
+    H, y = make_blobs(n_samples=50, random_state=0, n_features=64)
+    H = StandardScaler().fit_transform(H)
+    labels = HDBSCAN(
+        algorithm="auto",
+        metric="seuclidean",
+        metric_params={"V": np.ones(H.shape[1])},
+    ).fit_predict(H)
+    n_clusters = len(set(labels)) - int(-1 in labels)
+    assert n_clusters == n_clusters_true
+
+
+def test_hdbscan_best_balltree_metric():
+    labels = HDBSCAN(
+        metric="seuclidean", metric_params={"V": np.ones(X.shape[1])}
+    ).fit_predict(X)
+    n_clusters = len(set(labels)) - int(-1 in labels)
+    assert n_clusters == n_clusters_true
+
+
+def test_hdbscan_no_clusters():
+    labels = HDBSCAN(min_cluster_size=len(X) - 1).fit_predict(X)
+    n_clusters = len(set(labels)) - int(-1 in labels)
+    assert n_clusters == 0
+
+
+def test_hdbscan_min_cluster_size():
+    """
+    Test that the smallest non-noise cluster has at least `min_cluster_size`
+    many points
+    """
+    for min_cluster_size in range(2, len(X), 1):
+        labels = HDBSCAN(min_cluster_size=min_cluster_size).fit_predict(X)
+        true_labels = [label for label in labels if label != -1]
+        if len(true_labels) != 0:
+            assert np.min(np.bincount(true_labels)) >= min_cluster_size
+
+
+def test_hdbscan_callable_metric():
+    metric = distance.euclidean
+    labels = HDBSCAN(metric=metric).fit_predict(X)
+    n_clusters = len(set(labels)) - int(-1 in labels)
+    assert n_clusters == n_clusters_true
+
+
+def test_hdbscan_input_lists():
```

Comment:
Can we parameterize this with `test_hdbscan_precomputed_array_like`?

Reply:
The tests are somewhat separate in intent, and parametrizing them together would be a cross product of `(list_with_no_inf, list_with_inf) x (precomputed, euclidean)` where `list_with_inf and euclidean` would be skipped.

Unless you had something particular in mind, or would still prefer that.

Reply:
I was not thinking of the cross product, but more of separate kwarg + input pairs:

```python
@pytest.mark.parameterize("kwargs, input", [
    ({"min_samples": 1}, [[1.0, 2.0], ...]),
    ({"metric": "precomputed"}, [...]),
])
```

For me, I see them both as "smoke tests" because they are checking something runs, but not checking if it is doing something reasonable.